legend-depot-core-services/src/main/java/org/finos/legend/depot/services/projects/ProjectsServiceImpl.java

```diff
@@ -137,11 +139,18 @@ public Set<ProjectVersionDependencies> getDependencyTree(List<ProjectVersion> pr
             fullDependencies.add(projectVersionDependencyTree);
             String fullPath = (parentPath == null ? "" : parentPath + PATH_DELIMITER) + projectVersionDependencyTree.getGav();
             projectVersionDependencyTree.setPath(fullPath);
-            ProjectData projectData = getProject(projectVersion.getGroupId(), projectVersion.getArtifactId());
+            String projectCoordinates = projectVersion.getGroupId() + projectVersion.getArtifactId();
+            if (!projectDataMap.containsKey(projectCoordinates))
+            {
+                // only fetch project if we haven't fetched it already
+                ProjectData project = getProject(projectVersion.getGroupId(), projectVersion.getArtifactId());
+                projectDataMap.put(projectCoordinates, project);
+            }
+            ProjectData projectData = projectDataMap.get(projectCoordinates);
```

Comment:
Probably can use `getIfAbsentPut`

Reply:
as discussed, can't use this as it would require us to get projectData from mongo which I am trying to avoid